kube-controllers/pkg/controllers/node/ipam.go

```diff
@@ -1049,6 +1240,114 @@ func (c *ipamController) kubernetesNodeForCalico(cnode string) (string, error) {
+
+	blocksGauges[poolName] = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name:        "ipam_blocks",
+		Help:        "IPAM blocks currently allocated for the IP pool",
+		ConstLabels: prometheus.Labels{"ippool": poolName},
+	}, []string{"node"})
+	prometheus.MustRegister(blocksGauges[poolName])
+
+	gcCandidateGauges[poolName] = prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "ipam_allocations_gc_candidates",
+		Help: "Allocations that are currently marked by the garbage collector as potential candidates to " +
+			"reclaim. Under normal operation, this metric should return to zero after the garbage collector " +
+			"confirms that this allocation can be reclaimed and reclaims it",
+		ConstLabels: prometheus.Labels{"ippool": poolName},
+	}, []string{"node"})
+	prometheus.MustRegister(gcCandidateGauges[poolName])
+
+	gcReclamationCounters[poolName] = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Name: "ipam_allocations_gc_reclamations",
+		Help: "The total allocations that have been reclaimed by the garbage collector over time. Under normal " +
+			"operation, this counter should increase, and increases of this counter should align to a return to zero " +
+			"for the candidate gauge.",
+		ConstLabels: prometheus.Labels{"ippool": poolName},
+	}, []string{"node"})
+	prometheus.MustRegister(gcReclamationCounters[poolName])
+}
+
+func (c *ipamController) unregisterMetricVectorsForPool(poolName string) {
+	if c.poolCache[poolName] == nil {
+		log.Warnf("Skipping deletion of pool metrics for unknown pool %s", poolName)
+		return
+	}
+
+	prometheus.Unregister(inUseAllocationGauges[poolName])
+	delete(inUseAllocationGauges, poolName)
+
+	prometheus.Unregister(borrowedAllocationGauges[poolName])
+	delete(borrowedAllocationGauges, poolName)
+
+	prometheus.Unregister(blocksGauges[poolName])
+	delete(blocksGauges, poolName)
+
+	prometheus.Unregister(gcCandidateGauges[poolName])
+	delete(gcCandidateGauges, poolName)
+
+	prometheus.Unregister(gcReclamationCounters[poolName])
+	delete(gcReclamationCounters, poolName)
+}
+
+func updatePoolGaugeWithNodeValues(gaugesByPool map[string]*prometheus.GaugeVec, pool string, nodeValues map[string]int) {
+	poolGauge := gaugesByPool[pool]
+	if poolGauge == nil {
+		log.Warnf("Gauge metric vector used for pool %s was not created, skipping publishing", pool)
+		return
+	}
+
+	poolGauge.Reset()
+	for node, value := range nodeValues {
+		poolGauge.With(prometheus.Labels{"node": node}).Set(float64(value))
+	}
+}
+
+func publishPoolSizeMetric(pool *apiv3.IPPool, poolNet *cnet.IPNet) {
+	ones, bits := poolNet.Mask.Size()
+	poolSize := math.Pow(2, float64(bits-ones))
+	poolSizeGauge.With(prometheus.Labels{"ippool": pool.Name}).Set(poolSize)
+}
+
+func clearPoolSizeMetric(poolName string) {
+	poolSizeGauge.Delete(prometheus.Labels{"ippool": poolName})
+}
+
+// When we stop tracking a node, clear counters to prevent accumulation of stale metrics.
+func clearNodeFromPoolCounters(node string) {
```

Comment:
This function name is a little confusing, because it seems to only clear from the reclamation counters and not any others. I think this is because it's the only metric that has both node and pool as labels? 

Might want to rename this to "clearReclaimedIPCountForNode" to be more clear. Just a minor thing.

Reply:
Agreed, I prefer the more specific naming you've suggested. When I chose this name, it was intended to be symmetric with the other metric functions, which at the time were all generic (e.g. operating on all metrics / all gauges). Now that there are metric-specific functions there's no upside to this naming choice

Reply:
And also just to clarify, the reclamation counter is the only metric to have an explicit clear function since all of the other gauge metrics are cleared (in `updateMetrics`) as part of calculation and republishing. This ensures any nodes that are no longer contributing metrics are cleared